source/common/conn_pool/conn_pool_base.cc

```diff
@@ -19,13 +19,45 @@ namespace {
 }
 } // namespace
 
+Network::TransportSocketOptionsConstSharedPtr
+wrapTransportSocketOptions(Network::TransportSocketOptionsConstSharedPtr transport_socket_options,
+                           Upstream::HostConstSharedPtr host) {
+  const auto& upstream_http_protocol_options = host->cluster().upstreamHttpProtocolOptions();
+  if (upstream_http_protocol_options && upstream_http_protocol_options->auto_sni_from_upstream()) {
+    const absl::string_view hostname = host->hostname();
+    if (!hostname.empty()) {
+      Network::TransportSocketOptionsConstSharedPtr options_with_decorated_server_name;
+      if (transport_socket_options) {
+        options_with_decorated_server_name =
+            std::make_shared<Network::ServerNameDecoratingTransportSocketOptions>(
+                hostname, transport_socket_options);
+      } else {
+        options_with_decorated_server_name = std::make_shared<Network::TransportSocketOptionsImpl>(
+            hostname, std::vector<std::string>{}, std::vector<std::string>{},
+            std::vector<std::string>{});
+      }
+      if (upstream_http_protocol_options->auto_san_validation()) {
+        return std::make_shared<Network::SubjectAltNamesDecoratingTransportSocketOptions>(
+            std::vector<std::string>{std::string(hostname)}, options_with_decorated_server_name);
+      } else {
+        return options_with_decorated_server_name;
+      }
+    } else {
+      return transport_socket_options;
+    }
+  } else {
+    return transport_socket_options;
+  }
+}
```

Comment:
nit: can you refactor this to early return from some of the if statements instead of having so much nesting? I think it should be pretty easy.

Reply:
Addressed